src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,463 @@
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "../libs/decoders/dr_flac.h"
+#include "channel_names.h"
+#include "checks.h"
+#include "mixer.h"
+#include "setup.h"
+#include "timer.h"
+
+CHECK_NARROWING();
+
+static std::unique_ptr<DiskNoises> disk_noises;
+static const unsigned int DiskNoiseSampleRateInHz = 22050;
+
+DiskNoises::DiskNoises(const bool enable_floppy_disk_noise,
+                       const bool enable_hard_disk_noise,
+                       const std::string& spin_up, const std::string& spin,
+                       const std::vector<std::string>& hdd_seek_samples,
+                       const std::string& floppy_spin_up,
+                       const std::string& floppy_spin,
+                       const std::vector<std::string>& floppy_seek_samples)
+{
+	MIXER_LockMixerThread();
+	const auto mixer_callback = std::bind(&DiskNoises::AudioCallback,
+	                                      this,
+	                                      std::placeholders::_1);
+	mix_channel               = MIXER_AddChannel(mixer_callback,
+                                       DiskNoiseSampleRateInHz,
+                                       ChannelName::DiskNoise,
+	                                             {ChannelFeature::Stereo});
+	mix_channel->Enable(true);
+	MIXER_UnlockMixerThread();
+	float vol_gain = percentage_to_gain(100);
+	mix_channel->SetAppVolume({vol_gain, vol_gain});
+
+	hdd_noise = std::make_unique<DiskNoiseDevice>(DiskType::HardDisk,
+	                                              enable_hard_disk_noise,
+	                                              spin_up,
+	                                              spin,
+	                                              hdd_seek_samples,
+	                                              true);
+	active_devices.push_back(hdd_noise.get());
+
+	floppy_noise = std::make_unique<DiskNoiseDevice>(DiskType::Floppy,
+	                                                 enable_floppy_disk_noise,
+	                                                 floppy_spin_up,
+	                                                 floppy_spin,
+	                                                 floppy_seek_samples,
+	                                                 false);
+	active_devices.push_back(floppy_noise.get());
+}
+
+void DiskNoises::AudioCallback(const int num_frames_requested)
+{
+	// stereo interleaved buffer
+	static std::vector<AudioFrame> out = {};
+	out.clear();
+
+	// Mix audio frames from all active devices
+	for (int i = 0; i < num_frames_requested; ++i) {
+		AudioFrame mixed_sample = {0.0f, 0.0f};
+		for (auto* device : disk_noises->active_devices) {
+			AudioFrame sample = device->GetNextFrame();
+			mixed_sample += sample;
+		}
+		out.emplace_back(mixed_sample);
+	}
+
+	disk_noises->mix_channel->AddAudioFrames(out);
+}
+
+DiskNoises::~DiskNoises()
+{
+	if (floppy_noise) {
+		floppy_noise.reset();
+	}
+
+	if (hdd_noise) {
+		hdd_noise.reset();
+	}
+	active_devices.clear();
```

Comment:
Another race condition here. The audio callback could still be running and you've just cleared `active_devices` which the callback could be iterating over.

I think I would just wrap this entire function in a `MIXER_LockMixerThread()` + `MIXER_UnlockMixerThread()` pair to be safe.